File: substrate/recursive_twin_bind.py
"""Recursive twin bind for every information asset (pure, fail-closed).

Every information asset may carry a twin of insights and questions.
This pure layer decides whether a twin *link* may be proposed and carries
only operator- or LLM-provided notes — never invents content from asset text.

twin_created is always False here (bind decision / payload only).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

TwinBindSource = Literal["operator", "llm_note_taker", "highlight_seed", "unknown"]

MAX_ID = 256
MAX_NOTE = 4000


class RecursiveTwinBindError(ValueError):
    """Fail-closed validation for recursive twin bind."""


@dataclass(frozen=True)
class RecursiveTwinBindDecision:
    parent_asset_id: str
    twin_id: str | None
    bind_allowed: bool
    twin_created: bool
    insights: tuple[str, ...]
    questions: tuple[str, ...]
    source: TwinBindSource
    llm_filled: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "parent_asset_id": self.parent_asset_id,
            "twin_id": self.twin_id,
            "bind_allowed": self.bind_allowed,
            "twin_created": False,
            "insights": list(self.insights),
            "questions": list(self.questions),
            "source": self.source,
            "llm_filled": self.llm_filled,
            "notes": list(self.notes),
            "authority": "twin_bind_advisory",
        }


def _require_bool(value: object, *, field: str) -> bool:
    if not isinstance(value, bool):
        raise RecursiveTwinBindError(f"{field} must be an explicit boolean")
    return value


def _require_nonempty(value: object, *, field: str, max_len: int = MAX_ID) -> str:
    if not isinstance(value, str):
        raise RecursiveTwinBindError(f"{field} must be a string")
    text = value.strip()
    if not text:
        raise RecursiveTwinBindError(f"{field} must be non-empty")
    if len(text) > max_len:
        raise RecursiveTwinBindError(f"{field} exceeds {max_len} chars")
    return text


def _clean_string_list(value: object | None, *, field: str) -> tuple[str, ...]:
    if value is None:
        return ()
    if not isinstance(value, list):
        raise RecursiveTwinBindError(f"{field} must be a list of strings or null")
    out: list[str] = []
    for i, item in enumerate(value):
        if not isinstance(item, str):
            raise RecursiveTwinBindError(f"{field}[{i}] must be a string")
        t = item.strip()
        if not t:
            continue
        if len(t) > MAX_NOTE:
            raise RecursiveTwinBindError(f"{field}[{i}] exceeds {MAX_NOTE} chars")
        out.append(t)
    return tuple(out)


def evaluate_recursive_twin_bind(
    *,
    parent_asset_id: object,
    gated: object,
    llm_filled: object,
    source: object,
    twin_id: object | None = None,
    insights: object | None = None,
    questions: object | None = None,
) -> RecursiveTwinBindDecision:
    """Decide whether a recursive twin bind may proceed. Never invents content."""
    gated_b = _require_bool(gated, field="gated")
    llm_filled_b = _require_bool(llm_filled, field="llm_filled")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")

    if not isinstance(source, str) or source not in (
        "operator",
        "llm_note_taker",
        "highlight_seed",
        "unknown",
    ):
        raise RecursiveTwinBindError(
            "source must be operator|llm_note_taker|highlight_seed|unknown"
        )
    src: TwinBindSource = source  # type: ignore[assignment]

    tid: str | None = None
    if twin_id is not None:
        if not isinstance(twin_id, str) or not twin_id.strip():
            raise RecursiveTwinBindError("twin_id must be non-empty string or null")
        tid = twin_id.strip()
        if len(tid) > MAX_ID:
            raise RecursiveTwinBindError(f"twin_id exceeds {MAX_ID} chars")

    notes: list[str] = []

    if gated_b is True:
        notes.append("gated parent asset — bind_allowed=false")
        return RecursiveTwinBindDecision(
            parent_asset_id=parent,
            twin_id=tid,
            bind_allowed=False,
            twin_created=False,
            insights=(),
            questions=(),
            source=src,
            llm_filled=False,
            notes=tuple(notes + ["twin_created=false", "insights/questions withheld"]),
            authority="twin_bind_advisory",
        )

    insight_t = _clean_string_list(insights, field="insights")
    question_t = _clean_string_list(questions, field="questions")

    if llm_filled_b is True:
        if src != "llm_note_taker":
            raise RecursiveTwinBindError(
                "llm_filled=true requires source=llm_note_taker (no invent provenance)"
            )
        if not insight_t and not question_t:
            raise RecursiveTwinBindError(
                "llm_filled=true requires non-empty insights or questions from caller (no invent)"
            )
        notes.append("llm_note_taker payload accepted — content is caller-supplied only")
    else:
        if src == "llm_note_taker":
            raise RecursiveTwinBindError(
                "source=llm_note_taker requires llm_filled=true with supplied lists"
            )
        if not insight_t and not question_t:
            notes.append(
                "no insights/questions supplied — bind still allowed as empty twin scaffold"
            )
        else:
            notes.append("operator/highlight insights-questions accepted as-supplied")

    if src == "unknown":
        notes.append("source=unknown — bind_allowed=false (provenance required)")
        return RecursiveTwinBindDecision(
            parent_asset_id=parent,
            twin_id=tid,
            bind_allowed=False,
            twin_created=False,
            insights=(),
            questions=(),
            source=src,
            llm_filled=False,
            notes=tuple(notes + ["twin_created=false"]),
            authority="twin_bind_advisory",
        )

    notes.append("bind_allowed=true — pure decision only")
    notes.append("twin_created=false")
    return RecursiveTwinBindDecision(
        parent_asset_id=parent,
        twin_id=tid,
        bind_allowed=True,
        twin_created=False,
        insights=insight_t,
        questions=question_t,
        source=src,
        llm_filled=llm_filled_b,
        notes=tuple(notes),
        authority="twin_bind_advisory",
    )


__all__ = [
    "RecursiveTwinBindDecision",
    "RecursiveTwinBindError",
    "evaluate_recursive_twin_bind",
]

File: apps/reading/src/api/recursiveTwinBind.ts
/**
 * Recursive twin bind for every information asset (pure client).
 *
 * Operator vision: every information asset has a twin substrate of insights
 * and questions. This pure layer decides whether a twin *link* may be
 * proposed and carries only operator- or LLM-provided notes — never invents
 * content from asset text.
 *
 * twin_created is always false here (bind decision / payload only).
 */

export type TwinBindSource =
  | "operator"
  | "llm_note_taker"
  | "highlight_seed"
  | "unknown";

export interface RecursiveTwinBindInput {
  parent_asset_id: string;
  /** Optional existing twin id (rebind / ensure). */
  twin_id?: string | null;
  insights?: string[] | null;
  questions?: string[] | null;
  /**
   * Provenance of insights/questions.
   * llm_note_taker requires llm_filled=true with non-empty lists from caller.
   */
  source: TwinBindSource;
  /** Explicit: did an LLM fill insights/questions? Never invent true. */
  llm_filled: boolean;
  /** Explicit gate on parent asset body. */
  gated: boolean;
}

export interface RecursiveTwinBindDecision {
  parent_asset_id: string;
  twin_id: string | null;
  bind_allowed: boolean;
  /** Always false in pure layer — no twin store write. */
  twin_created: boolean;
  insights: string[];
  questions: string[];
  source: TwinBindSource;
  llm_filled: boolean;
  notes: string[];
  authority: "twin_bind_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function cleanStringList(value: unknown, name: string): string[] {
  if (value == null) return [];
  if (!Array.isArray(value)) {
    throw new Error(`${name} must be an array of strings or null`);
  }
  return value.map((item, i) => {
    if (typeof item !== "string") {
      throw new Error(`${name}[${i}] must be a string`);
    }
    return item.trim();
  }).filter(Boolean);
}

/**
 * Decide whether a recursive twin bind may proceed for an asset.
 * Never invents insights/questions from asset text.
 * twin_created always false.
 */
export function evaluateRecursiveTwinBind(
  input: RecursiveTwinBindInput,
): RecursiveTwinBindDecision {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.gated !== "boolean") {
    throw new Error(
      "gated must be an explicit boolean from asset provenance (fail closed)",
    );
  }
  if (typeof input.llm_filled !== "boolean") {
    throw new Error("llm_filled must be an explicit boolean (fail closed)");
  }
  const parent = requireNonEmpty(input.parent_asset_id, "parent_asset_id");
  const source = input.source;
  if (
    source !== "operator" &&
    source !== "llm_note_taker" &&
    source !== "highlight_seed" &&
    source !== "unknown"
  ) {
    throw new Error("source must be operator|llm_note_taker|highlight_seed|unknown");
  }

  const notes: string[] = [];
  let twinId: string | null = null;
  if (input.twin_id != null && input.twin_id !== undefined) {
    if (typeof input.twin_id !== "string" || !input.twin_id.trim()) {
      throw new Error("twin_id must be non-empty string or null");
    }
    twinId = input.twin_id.trim();
  }

  if (input.gated === true) {
    notes.push("gated parent asset — bind_allowed=false");
    return {
      parent_asset_id: parent,
      twin_id: twinId,
      bind_allowed: false,
      twin_created: false,
      insights: [],
      questions: [],
      source,
      llm_filled: false,
      notes: [...notes, "twin_created=false", "insights/questions withheld"],
      authority: "twin_bind_advisory",
    };
  }

  const insights = cleanStringList(input.insights, "insights");
  const questions = cleanStringList(input.questions, "questions");

  if (input.llm_filled === true) {
    if (source !== "llm_note_taker") {
      throw new Error(
        "llm_filled=true requires source=llm_note_taker (no invent provenance)",
      );
    }
    if (insights.length === 0 && questions.length === 0) {
      throw new Error(
        "llm_filled=true requires non-empty insights or questions from caller (no invent)",
      );
    }
    notes.push("llm_note_taker payload accepted — content is caller-supplied only");
  } else {
    if (source === "llm_note_taker") {
      throw new Error(
        "source=llm_note_taker requires llm_filled=true with supplied lists",
      );
    }
    if (insights.length === 0 && questions.length === 0) {
      notes.push(
        "no insights/questions supplied — bind still allowed as empty twin scaffold",
      );
    } else {
      notes.push("operator/highlight insights-questions accepted as-supplied");
    }
  }

  if (source === "unknown") {
    notes.push("source=unknown — bind_allowed=false (provenance required)");
    return {
      parent_asset_id: parent,
      twin_id: twinId,
      bind_allowed: false,
      twin_created: false,
      insights: [],
      questions: [],
      source,
      llm_filled: false,
      notes: [...notes, "twin_created=false"],
      authority: "twin_bind_advisory",
    };
  }

  notes.push("bind_allowed=true — pure decision only");
  notes.push("twin_created=false");

  return {
    parent_asset_id: parent,
    twin_id: twinId,
    bind_allowed: true,
    twin_created: false,
    insights,
    questions,
    source,
    llm_filled: input.llm_filled,
    notes,
    authority: "twin_bind_advisory",
  };
}

export function formatTwinBindSummary(d: RecursiveTwinBindDecision): string {
  return (
    `parent=${d.parent_asset_id} · bind_allowed=${d.bind_allowed} · ` +
    `twin_created=false · insights=${d.insights.length} · questions=${d.questions.length}`
  );
}
